pcdsdevices/lodcm.py

```diff
@@ -60,6 +69,437 @@ def __init__(self, prefix, *args, **kwargs):
+    h1n = Cpt(IMS, ':MON:MMS:09', kind='normal', doc='LOM Xtal1 Hn')
+    # paralell to the crystal surface movement
+    h1p = Cpt(IMS, ':MON:MMS:20', kind='normal', doc='LOM Xtal1 Hp')
+
+    tab_component_names = True
+
+    def __init__(self, prefix, *args, **kwargs):
+        super().__init__(prefix, *args, **kwargs)
+
+
+class CrystalTower2(Device):
+    """
+    Crystal Tower 2.
+
+    Parameters
+    ----------
+    prefix : str
+        Epics base Pv prefix.
+    """
+    # x, y, and z are on the base but not touched in normal operations
+    z2 = Cpt(IMS, ':MON:MMS:10', kind='normal', doc='LOM Xtal2 Z')
+    x2 = Cpt(IMS, ':MON:MMS:11', kind='normal', doc='LOM Xtal2 X')
+    y2 = Cpt(IMS, ':MON:MMS:12', kind='normal', doc='LOM Xtal2 Y')
+    # thata movement
+    th2 = Cpt(IMS, ':MON:MMS:13', kind='normal', doc='LOM Xtal2 Theta')
+    # chi movement
+    ch2 = Cpt(IMS, ':MON:MMS:14', kind='normal', doc='LOM Xtal2 Chi')
+    # normal to the crystal surface movement
+    h2n = Cpt(IMS, ':MON:MMS:15', kind='normal', doc='LOM Xtal2 Hn')
+    # in the DAQ for scanning in python, only used for commissioning
+    diode2 = Cpt(IMS, ':MON:MMS:21', kind='normal', doc='LOM Xtal2 PIPS')
+
+    tab_component_names = True
+
+    def __init__(self, prefix, *args, **kwargs):
+        super().__init__(prefix, *args, **kwargs)
+
+
+class DiagnosticsTower(Device):
+    """
+    Diagnostic Tower Motors.
+
+    Parameters
+    ----------
+    prefix : str
+        Epics base PV prefix.
+    """
+    # Located midway between T1 and T2 in the center of rotation of the device.
+    # horizontal slits
+    dh = Cpt(IMS, ':MON:MMS:16', kind='normal', doc='LOM Dia H')
+    # vertical slits
+    dv = Cpt(IMS, ':MON:MMS:17', kind='normal', doc='LOM Dia V')
+    dr = Cpt(IMS, ':MON:MMS:19', kind='normal', doc='LOM Dia Theta')
+    # filters wheel
+    df = FCpt(IMS, '{self._prefix}:MON:MMS:{self._df_suffix}', kind='normal',
+              doc='LOM Dia Filter Wheel')
+    # pips diode
+    dd = Cpt(IMS, ':MON:MMS:18', kind='normal', doc='LOM Dia PIPS')
+    # yag screen
+    yag_zoom = Cpt(IMS, ':MON:CLZ:01', kind='normal', doc='LOM Zoom')
+
+    tab_component_names = True
+
+    def __init__(self, prefix, *args, **kwargs):
+        self._prefix = prefix
+        self._df_suffix = '27'
+        if 'XCS' in self._prefix:
+            self._df_suffix = '22'
+        super().__init__(prefix, *args, **kwargs)
+
+
+class OffsetIMS(PseudoSingleInterface):
+    # offset_pv = Cpt(EpicsSignal, kind='omitted')
+
+    # def move(pos):
+    #     self.mv(pos + Pv.get(offset_pv))
+
+    # def wm():
+    #     return motor.wm() - Pv.get(offset_pv)
+
+    # def set(value):
+    #     if use_ims_preset:
+    #         Pv.put("%s_SET" % offset_pv, motor.wm() - value)
+    #     Pv.put(offset_pv, motor.wm() - value)
+
+    def __init__(self, prefix, *args, **kwargs):
+        super().__init__('', *args, **kwargs)
+
+
+class CrystalTower1State(BaseInterface, Device):
```

Comment:
So I separated the 'state' related functions into separated classes, because I'm using them in the calculations, and since I did not need all the motors there I did not want to include them here, but I think I should include the motors here and have only 2 classes (Tower1, Tower2) instead of 4 (Tower1motors, Tower1States, .....)

Reply:
In the calculations I only use `th1` motor.....

Reply:
I think it's OK if the calculation class has some redundancy... this mega device would be very hard to write without it. I wish there was an easy way to have something like a "reference" component, e.g. a tool like:
```
class Energy(PseudoPositioner):
    energy = Cpt(PseudoSingleInterface)
    th1 = RefCpt('{parent.tower1.th1}')
```
Would be neat if it existed/were possible in the `ophyd` framework

Reply:
Yes, this would be very nice, I agree